src/artifact-uploader.cpp

```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright (C) 2017 Prism Skylabs
+ */
+#include "artifact-uploader.h"
+#include "private/UploadQueue.h"
+#include "private/OutputController.h"
+#include "private/UploadTaskQueuer.h"
+#include "private/util.h"
+#include "boost/make_shared.hpp"
+#include "private/const-strings.h"
+#include "easylogging++.h"
+#include "boost/filesystem.hpp"
+
+namespace prism
+{
+namespace connect
+{
+
+class ArtifactUploader::Impl
+{
+public:
+    Impl();
+    ~Impl();
+
+    Status init(const ArtifactUploader::Configuration& cfg,
+                ArtifactUploader::ClientConfigCallback* configCallback);
+
+//    ~ArtifactUploader();
+
+    // all upload* methods are asynchronous, non-blocking, take ownership
+    // of data passed to them
+    void uploadBackground(const timestamp_t& timestamp, PayloadAuPtr payload);
+    void uploadObjectStream(const ObjectStream& stream, PayloadAuPtr payload);
+    void uploadFlipbook(const Flipbook& flipbook, PayloadAuPtr payload);
+    void uploadEvent(const timestamp_t& timestamp, move_ref<Events> data);
+
+private:
+    UploadQueuePtr uploadQueue_;
+    OutputControllerPtr outputController_;
+    UploadTaskQueuerPtr uploadTaskQueuer_;
+};
+
+ArtifactUploader::ArtifactUploader()
+    : pImpl_(new Impl())
+{
+
+}
+
+Status ArtifactUploader::init(const ArtifactUploader::Configuration& cfg,
+                              ArtifactUploader::ClientConfigCallback* configCallback)
+{
+    return pImpl_->init(cfg, configCallback);
+}
+
+void ArtifactUploader::uploadBackground(const timestamp_t& timestamp, PayloadAuPtr payload)
+{
+    pImpl_->uploadBackground(timestamp, payload);
+}
+
+void ArtifactUploader::uploadObjectStream(const ObjectStream& stream, PayloadAuPtr payload)
+{
+    pImpl_->uploadObjectStream(stream, payload);
+}
+
+void ArtifactUploader::uploadFlipbook(const Flipbook& flipbook, PayloadAuPtr payload)
+{
+    pImpl_->uploadFlipbook(flipbook, payload);
+}
+
+void ArtifactUploader::uploadEvent(const timestamp_t& timestamp, move_ref<Events> events)
+{
+    pImpl_->uploadEvent(timestamp, events);
+}
+
+ArtifactUploader::Impl::~Impl()
+{
+    if (uploadTaskQueuer_)
+        uploadTaskQueuer_->finalizeUpload();
+
+    if (outputController_)
+        outputController_->stop();
+}
+
+Status ArtifactUploader::Impl::init(const ArtifactUploader::Configuration& cfg,
+                                    ArtifactUploader::ClientConfigCallback* configCallback)
+{
+    if (cfg.maxQueueSizeMB <= 1e-5)
+    {
+        LOG(ERROR) << "Invalid maxQueueSizeMB value " << cfg.maxQueueSizeMB;
+        return makeError();
+    }
+
+    if (cfg.warnQueueSizeMB <= 1e-5)
```

Comment:
It's a question to DenisB about floating type.
YuriyS, how to do you suggest to check values? One way would be to calculate integer value in bytes and perform check on that value.